src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java

```diff
@@ -563,17 +585,47 @@ private Path prepareAndroidManifest() throws IOException {
         Path targetSourcePath = paths.getSourcePath().resolve(targetOS);
 
         Path userManifest = targetSourcePath.resolve(Constants.MANIFEST_FILE);
+        ReleaseConfiguration releaseConfiguration = projectConfiguration.getReleaseConfiguration();
+
         if (!Files.exists(userManifest)) {
             // copy manifest to gensrc/android
             Path androidPath = paths.getGenPath().resolve(targetOS);
             Path genManifest = androidPath.resolve(Constants.MANIFEST_FILE);
             Logger.logDebug("Copy " + Constants.MANIFEST_FILE + " to " + genManifest.toString());
             FileOps.copyResource("/native/android/AndroidManifest.xml", genManifest);
             FileOps.replaceInFile(genManifest, "package='com.gluonhq.helloandroid'", "package='" + getAndroidPackageName() + "'");
-            FileOps.replaceInFile(genManifest, "A HelloGraal", projectConfiguration.getAppName());
+            String newAppLabel = Optional.ofNullable(releaseConfiguration.getAppLabel())
+                    .orElse(projectConfiguration.getAppName());
+            FileOps.replaceInFile(genManifest, "A HelloGraal", newAppLabel);
+            String newVersionCode = Optional.ofNullable(releaseConfiguration.getVersionCode())
+                    .orElse(DEFAULT_CODE_VERSION);
+            FileOps.replaceInFile(genManifest, ":versionCode='1'", ":versionCode='" + newVersionCode + "'");
+            String newVersionName = Optional.ofNullable(releaseConfiguration.getVersionName())
+                    .orElse(DEFAULT_CODE_NAME);
+            FileOps.replaceInFile(genManifest, ":versionName='1.0'", ":versionName='" + newVersionName + "'");
             Logger.logInfo("Default Android manifest generated in " + genManifest.toString() + ".\n" +
                     "Consider copying it to " + targetSourcePath.toString() + " before performing any modification");
             return androidPath;
+        } else {
+            // update manifest in src/android
+            String versionCode = FileOps.getNodeValue(userManifest.toString(), "manifest", ":versionCode");
```

Comment:
The `versionCode` variable might be `null`, in which case the replacement will not work as expected when `releaseConfiguration.getVersionCode()` is not empty. I would remove the usage of `Optional` and replace the if check like this:

    String versionCode = FileOps.getNodeValue(userManifest.toString(), "manifest", ":versionCode");
    String newVersionCode = releaseConfiguration.getVersionCode();
    if (versionCode != null && newVersionCode != null) {

The same pattern can be used on the other two replacements.

Reply:
makes sense, done